docs/concepts/time-and-windows.rst

```diff
@@ -0,0 +1,118 @@
+
+Finally, whenever a KSQL application writes records to Kafka,
+then it will also assign timestamps to these new records.
+The way the timestamps are assigned depends on the context:
+
+* When new output records are generated via directly processing some input record,
+  output record timestamps are inherited from input record timestamps directly.
+* When new output records are generated via periodic functions, the output record
+  timestamp is defined as the current internal time of the stream task.
+* For aggregations, the timestamp of the resulting update record will be that of
+  the latest input record that triggered the update.
+
+.. tip::
+
+Know your time: When working with time you should also make sure that additional
+aspects of time such as time zones and calendars are correctly synchronized – or
+at least understood and traced – throughout your stream data pipelines. It often
+helps, for example, to agree on specifying time information in UTC or in Unix time
+(such as seconds since the epoch). You should also not mix topics with different time semantics.
+
+
+Windowing
+*********
+
+Windowing lets you control how to group records that have the same key for stateful
+operations such as aggregations or joins into so-called windows. Windows are tracked
+per record key.
+
+When working with windows, you can specify a retention period for the window.
+This retention period controls how long KSQL waits for out-of-order or late-arriving records
+for a given window. If a record arrives after the retention period of a window has passed,
+the record is discarded and will not be processed in that window.
+
+We added a new configuration that we call "grace period" that determines how long
+we wait before we close a window. Retention time is still a valid parameter that defines
+how long we store the (potentially) closed window -- we do this to allow to access the
+window via Interactive Queries even if it's already closed.
+
+Late-arriving records are always possible in the real world and should be properly accounted
+for in your applications. It depends on the effective time semantics how late records are handled.
+In the case of processing-time, the semantics are “when the record is being processed”, which means
+that the notion of late records is not applicable as, by definition, no record can be late.
+Hence, late-arriving records can only be considered as such (i.e. as arriving “late”) for
+event-time or ingestion-time semantics. In both cases, KSQL is able to properly
+handle late-arriving records.
+
+.. image:: ../img/ksql-window.png
+
+
+.. image:: ../img/ksql-window-aggregation.png
+
+
+Session Window start and end time are both inclusive (in contrast to time-windows),
+and there is always a record in the session window with the start and end timestamps
+(because the timestamp of the first and last record in the window define window start and end time).

```

Comment:
Would be great to add a new section around windowed joins, using the `WITHIN` clause, e.g. 

```
SELECT O.ORDER_ID, O.TOTAL_AMOUNT, O.CUSTOMER_NAME,
          S.SHIPMENT_ID, S.WAREHOUSE
          FROM NEW_ORDERS O
          INNER JOIN SHIPMENTS S
            WITHIN 1 HOURS
            ON O.ORDER_ID = S.ORDER_ID;
```